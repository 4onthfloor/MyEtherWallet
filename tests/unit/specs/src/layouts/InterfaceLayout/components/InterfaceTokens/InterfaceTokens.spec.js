import Vue from 'vue';
import sinon from 'sinon';
import { shallowMount } from '@vue/test-utils';
import InterfaceTokens from '@/layouts/InterfaceLayout/components/InterfaceTokens/InterfaceTokens.vue';
import InterfaceTokensModal from '@/layouts/InterfaceLayout/components/InterfaceTokensModal/InterfaceTokensModal.vue';
import { Tooling } from '@@/helpers';

const showModal = sinon.stub();
const hideModal = sinon.stub();
const BModalStub = {
  name: 'b-modal',
  template: '<div><slot></slot></div>',
  props: ['to'],
  methods: {
    show: showModal,
    hide: hideModal
  }
};

describe('InterfaceTokens.vue', () => {
  const customTokens = [
    { name: 'ETH', balance: 200 },
    { name: 'BTH', balance: 300 },
    { name: 'ETC', balance: 400 }
  ];

  const localTokens = [
    { name: 'ETH', balance: 200 },
    { name: 'BTH', balance: 300 },
    { name: 'ETC', balance: 400 }
  ];

  let localVue, i18n, wrapper, store;

  beforeAll(() => {
    const baseSetup = Tooling.createLocalVueInstance();
    localVue = baseSetup.localVue;
    i18n = baseSetup.i18n;
    store = baseSetup.store;

    Vue.config.warnHandler = () => {};
    Vue.config.errorHandler = () => {};
  });

  beforeEach(() => {
    wrapper = shallowMount(InterfaceTokens, {
      localVue,
      i18n,
      store,
      stubs: {
        'interface-tokens-modal': InterfaceTokensModal,
        'b-modal': BModalStub
      }
    });
  });

  afterAll(() => {
    wrapper.destroy();
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct search data', async () => {
    const search = 'search';
    wrapper.setData({ search });
    await wrapper.vm.$nextTick();
    const input = wrapper.find('.search-block input').element.value;
    expect(input).toEqual(search);
  });

  it('should render correct customTokens data', () => {
    wrapper.setData({ customTokens });

    const tableElement = wrapper.findAll('.token-table-container table').at(0);
    const trElements = tableElement.findAll('tr');
    trElements.wrappers.forEach((trElement, i) => {
      expect(
        trElement
          .findAll('td')
          .at(0)
          .element.text()
      ).toEqual(customTokens[i].name);
      expect(
        trElement
          .findAll('td')
          .at(1)
          .element.text()
      ).toEqual(String(customTokens[i].balance));
    });
  });

  it('should render correct localTokens data', () => {
    wrapper.setData({ localTokens });
    const tableElement = wrapper.findAll('.token-table-container table').at(1);
    const trElements = tableElement.findAll('tr');
    trElements.wrappers.forEach((trElement, i) => {
      expect(
        trElement
          .findAll('td')
          .at(0)
          .element.text()
      ).toEqual(localTokens[i].name);
      expect(
        trElement
          .findAll('td')
          .at(1)
          .element.text()
      ).toEqual(String(localTokens[i].balance));
    });
  });

  describe('InterfaceTokens.vue Methods', () => {
    it('should render correct addTokenModal method', () => {
      wrapper.vm.addTokenModal();
      expect(showModal.called).toBe(true);
    });
  });

  it('should dismount properly', () => {
    expect(wrapper.destroy()).toBe(undefined);
  });
});
