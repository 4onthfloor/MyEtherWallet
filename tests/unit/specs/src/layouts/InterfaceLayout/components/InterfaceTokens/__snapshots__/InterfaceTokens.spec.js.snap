// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterfaceTokens.vue should render correctly 1`] = `
<div class="transaction-tokens">
  <div>
    <div title="Add Custom Tokens" hide-footer="" centered="" static="" lazy="" class="bootstrap-modal nopadding max-height-1">
      <form class="tokens-modal-body">
        <div><input placeholder="Token Contract Address" name="Address" type="text" class="custom-input-text-1" aria-required="true" aria-invalid="false"> <span class="error-message" style="display: none;">
          Invalid address given.
        </span> <input placeholder="Token Symbol" name="Symbol" type="text" class="custom-input-text-1"> <input placeholder="Decimals" name="Decimal" type="number" min="0" max="18" class="custom-input-text-1"> <span class="error-message" style="display: none;">
          Invalid Decimal. Decimal can only be between 0 and 18.
        </span></div>
        <div class="button-block"><button class="disabled save-button large-round-button-green-filled clickable">
            Save
          </button>
          <interface-bottom-text-stub link="https://kb.myetherwallet.com" linktext="Help Center" question="Don't know what it is?"></interface-bottom-text-stub>
        </div>
      </form>
    </div>
  </div>
  <div class="wrap">
    <div class="tokens-container">
      <div class="token-search">
        <div class="block-title">
          <div class="title-container">
            <h4>Tokens</h4> <img alt="" src="~@/assets/images/icons/change.svg">
          </div>
          <p>+ Custom Tokens</p>
        </div>
        <div class="search-block"><input placeholder="Search" autocomplete="off"> <i aria-hidden="true" class="fa fa-search"></i></div>
      </div>
      <div class="cant-load" style="display: none;">
        Can't load balances in offline mode
      </div>
      <div class="token-table-container">
        <table style="display: none;"></table>
        <table style="display: none;"></table>
        <div class="spinner-container"><i class="fa fa-spinner fa-spin"></i></div>
        <div class="spinner-container" style="display: none;">
          No tokens found :(
        </div>
      </div>
      <!---->
    </div>
    <interface-ads-stub></interface-ads-stub>
  </div>
</div>
`;
